### ISSUE TITLE
No warning/error Severity CS8602 Dereference of a possibly null reference for optional argument

### ISSUE THREAD

**Issue by ssavkovic** [dotnet/roslyn#52792]
Microsoft Visual Studio Enterprise 2019
Version 16.9.4 (.NET Core 3.1 and .NET 5)

Looks like some edge case is missed when checking possible null references of optional parameters. In the code below the value for `data` reference will be null by default.

Function like:
`static void BadUse([Optional] Data data) => Console.WriteLine($"Data: ${data.Count}");`

Expected Behavior:
Warning	CS8602	Dereference of a possibly null reference.

Actual Behavior:
No suggestion/warning/error and null reference exception thrown on execution.

Note that it works fine when you do add DefaultParameterValue(null) attribute like (we get warning/error):
`static void BadUse([Optional, DefaultParameterValue(null)] Data data) => Console.WriteLine($"Data: ${data.Count}");`

**Comment by RikkiGibson** [comment]
Fix should probably be made in here:

https://github.com/dotnet/roslyn/blob/cd1d0fec5ecf8e23725e95bcf3426b358baf2a59/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs#L260-L260

It's possible that `defaultValue` is `null` in this scenario. We might need to duplicate or share the logic here:

https://github.com/dotnet/roslyn/blob/5ac580fd39728a01e1ce3c683970f62a24b76548/src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs#L1373-L1386

Basically we produce these warnings by creating a BoundParameterEqualsValue which "assigns" the default value, then analyzing it in the nullable walker. We expect `warning CS8625: Cannot convert null literal to non-nullable reference type.` in this scenario.
